From 1ad77150db0144c62bf0c4e3d7ba79cf8160b099 Mon Sep 17 00:00:00 2001
From: jizhenlo <zhenlong.z.ji@intel.com>
Date: Wed, 25 Oct 2023 12:16:07 +0800
Subject: [PATCH 49/54] Support share_data partition fast erase

Only erase the first 4KB of the partition and set to 0.
fs_mgr will format the partiton if all 0 in first 4kB.

Tracked-On: OAM-112711
Signed-off-by: jizhenlo <zhenlong.z.ji@intel.com>
---
 libfastboot/flash.c | 17 +++++++++++------
 1 file changed, 11 insertions(+), 6 deletions(-)

diff --git a/libfastboot/flash.c b/libfastboot/flash.c
index af08bdc..2edd898 100644
--- a/libfastboot/flash.c
+++ b/libfastboot/flash.c
@@ -59,6 +59,7 @@
 static struct gpt_partition_interface gparti;
 static UINT64 cur_offset;
 static BOOLEAN userdata_erased = FALSE;
+static BOOLEAN share_data_erased = FALSE;
 BOOLEAN new_install_device = FALSE;
 
 #define part_start (gparti.part.starting_lba * gparti.bio->Media->BlockSize)
@@ -603,20 +604,22 @@ static EFI_STATUS fast_erase_part(const CHAR16 *label)
 EFI_STATUS erase_by_label(CHAR16 *label)
 {
 	EFI_STATUS ret;
+	BOOLEAN is_data = (!StrCmp(label, L"userdata") || !StrCmp(label, L"data"));
+	BOOLEAN is_share_data = !StrCmp(label, L"share_data");
 
 	/* userdata/data partition only need to be erased once during each boot */
-	if (!StrCmp(label, L"userdata") || !StrCmp(label, L"data")) {
-		if (userdata_erased) {
-			debug(L"userdata/data partition had already been erased. skip.");
+	if (is_data || is_share_data) {
+		if ((is_data && userdata_erased) || (is_share_data && share_data_erased)) {
+			debug(L"userdata/share_data partition had already been erased. skip.");
 			return EFI_SUCCESS;
 		}
 
 		if (new_install_device) {
-			debug(L"New install devcie, fast erase userdata/data partition");
+			debug(L"New install devcie, fast erase userdata/share_data partition");
 			return fast_erase_part(label);
 		} else {
 #ifndef USER
-			debug(L"fast erase userdata/data partition for userdebug build");
+			debug(L"fast erase userdata/share_data partition for userdebug build");
 			return fast_erase_part(label);
 #endif
 		}
@@ -635,8 +638,10 @@ EFI_STATUS erase_by_label(CHAR16 *label)
 	if (!CompareGuid(&gparti.part.type, &EfiPartTypeSystemPartitionGuid))
 		return gpt_refresh();
 
-	if (!StrCmp(label, L"userdata") || !StrCmp(label, L"data"))
+	if (is_data)
 		userdata_erased = TRUE;
+	if (is_share_data)
+		share_data_erased = TRUE;
 
 	return EFI_SUCCESS;
 }
-- 
2.25.1

